### PR TITLE
fix(Overlay): Fix types and overlay example

### LINKED ISSUE #4717: Overlay TypeScript definition is incomplete and untested
## Describe the bug

The `Overlay` component has an incomplete TypeScript definition and, as far as I can tell, it's not tested either.

## To Reproduce

1. Check out master (I used https://github.com/react-bootstrap/react-bootstrap/commit/68127cbe67338cdc24f83dc66f869f9bfc875aaa)
2. Apply a TypeScript test:

<details>
<summary>Patch</summary>

```diff
diff --git a/types/simple.test.tsx b/types/simple.test.tsx
index 1a04a92e..01befedc 100644
--- a/types/simple.test.tsx
+++ b/types/simple.test.tsx
@@ -24,6 +24,7 @@ import {
   Modal,
   Nav,
   NavDropdown,
+  Overlay,
   OverlayTrigger,
   Tooltip,
   Pagination,
@@ -325,6 +326,45 @@ import {
   </NavDropdown>
 </Nav>;
 
+<Overlay
+  container={<div />}
+  onEnter={() => {}}
+  onEntered={() => {}}
+  onEntering={() => {}}
+  onExit={() => {}}
+  onExited={() => {}}
+  onExiting={() => {}}
+  onHide={() => {}}
+  placement="bottom"
+  popperConfig={{}}
+  rootClose
+  rootCloseEvent="click"
+  show={true}
+  target={<div />}
+  transition={true}
+>
+  {({
+    arrowProps,
+    placement,
+    ref,
+    show,
+    style,
+  }) => {
+    return (
+      <Popover
+        arrowProps={arrowProps}
+        id="overlay-controlled-popover"
+        placement={placement}
+        ref={ref}
+        style={style}
+      >
+        <Popover.Title as="h3">Title</Popover.Title>
+        <Popover.Content>some content</Popover.Content>
+      </Popover>
+    );
+  }}
+</Overlay>;
+
 <OverlayTrigger
   placement="left"
   overlay={

```

</details>

3. Run `yarn dtslint`
4. See error (below)

<details>
<summary>TypeScript error</summary>

```
$ yarn dtslint
yarn run v1.19.1
$ dtslint types
Error: /Users/kumar/tmp/react-bootstrap/types/simple.test.tsx:329:1
ERROR: 329:1  expect  TypeScript@next compile error:
No overload matches this call.
  Overload 2 of 2, '(props: OverlayProps, context?: any): Overlay', gave the following error.
    Type 'Element' is not assignable to type 'Element | Node | Component<any, {}, any> | ((props: object) => ComponentOrElement) | undefined'.
      Type 'Element' is not assignable to type 'Component<any, {}, any>'.
  Overload 2 of 2, '(props: OverlayProps, context?: any): Overlay', gave the following error.
    Type 'Element' is not assignable to type 'Element | Node | Component<any, {}, any> | ((props: object) => ComponentOrElement) | undefined'.
      Type 'Element' is not assignable to type 'Component<any, {}, any>'.
ERROR: 347:5  expect  TypeScript@next compile error:
Binding element 'arrowProps' implicitly has an 'any' type.
ERROR: 348:5  expect  TypeScript@next compile error:
Binding element 'placement' implicitly has an 'any' type.
ERROR: 349:5  expect  TypeScript@next compile error:
Binding element 'ref' implicitly has an 'any' type.
ERROR: 350:5  expect  TypeScript@next compile error:
Binding element 'show' implicitly has an 'any' type.
ERROR: 351:5  expect  TypeScript@next compile error:
Binding element 'style' implicitly has an 'any' type.
```

</details>

## Reproducible Example

(follow steps above)

## Expected behavior

There should be no TypeScript errors.

## Environment (please complete the following information)

- Operating System: macOS
- Browser, Version: n/a
- React-Bootstrap Version: https://github.com/react-bootstrap/react-bootstrap/commit/68127cbe67338cdc24f83dc66f869f9bfc875aaa

## Additional context

The test I added uses all available props and a realistic (working) integration with Popover. It suggests that the following types are not defined correctly:

* The `children` prop callback (its function signature is not defined).
* The `container` and `target` props (they have conflicting resolutions)

### PR DESCRIPTION
**Pull request by kyletsang**
Resolves #4717

### CODE REVIEW

File: src/Overlay.tsx [hunk 1]
```diff
@@ -17,9 +17,13 @@ export interface OverlayInjectedProps {
   style: React.CSSProperties;
   'aria-labelledby'?: string;
 
-  arrowProps: { ref: any; style: object };
+  arrowProps: {
+    ref: React.RefCallback<HTMLElement>;
```
Comment (bpas247) [inline src/Overlay.tsx:21]: are we explicitly expecting this to be a ref callback? why can't it just be typed as any valid [ref](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/201779301ed80fd8914ccfb4183401fe05eb44d5/types/react/index.d.ts#L92)?

Comment (kyletsang) [reply src/Overlay.tsx:21]: This is typed to match that of upstream react-overlays

https://github.com/react-bootstrap/react-overlays/blob/212ffc2aa72756c4c9a2054bb3093050fe0db75b/src/Overlay.tsx#L41-L44

Comment (bpas247) [reply src/Overlay.tsx:21]: okay well I don't understand why the types for react-overlays is that way either 😛 it should be allowed to pass in any valid refs for that feature.

But this is something we'd have to fix in react-overlays first so for now it makes sense for them to be the same type.

Comment (kyletsang) [reply src/Overlay.tsx:21]: Agreed.  I was wondering about the ref typing as well, but didn't have enough time for a deep dive into RO today.

Comment (kyletsang) [reply src/Overlay.tsx:21]: @bpas247 

Double checked RO and the type is correct. It always passes a callback ref.

https://github.com/react-bootstrap/react-overlays/blob/212ffc2aa72756c4c9a2054bb3093050fe0db75b/src/Overlay.tsx#L62

https://github.com/react-bootstrap/react-overlays/blob/212ffc2aa72756c4c9a2054bb3093050fe0db75b/src/Overlay.tsx#L137-L140

**Pull request merged**